src/lightning/fabric/fabric.py

```diff
@@ -678,8 +683,17 @@ def save(self, path: Union[str, Path], state: Dict[str, Union[nn.Module, Optimiz
             path: A path to where the file(s) should be saved
             state: A dictionary with contents to be saved. If the dict contains modules or optimizers, their
                 state-dict will be retrieved and converted automatically.
+            filter: An optional filter that returns a boolean indicating whether the given parameter should be saved
+                (``True``) or filtered out (``False``). It can be a callable that will be applied to each key in
+                ``state`` or a dictionary of the same format as ``state`` mapping each keys to a callable.
         """
-        self._strategy.save_checkpoint(path=path, state=_unwrap_objects(state))
+        if filter is not None:
+            if not isinstance(filter, dict):
+                filter = {k: filter for k in state}
```

Comment:
The filter shouldn't apply to elements of the state, but the elements of the model. There should be an error if a filter is provided for a key that doesn't correspond to a nn.Module.

Reply:
> The filter shouldn't apply to elements of the state, but the elements of the model

This is just input validation, it's applied to the resulting state_dicts in the actual implementation

> There should be an error if a filter is provided for a key that doesn't correspond to a nn.Module.

Why impose this limitation?

Reply:
The filtering is inclusive, meaning that I have to return True in the function if I want to include the contents and otherwise False. It would be very cumbersome or error prone to have to update the filter everytime I add new things to the state (e.g. my code evolveds and now I want to also save additional metadata). 

Please note that the whole point of #17724 is that currently, there is no way to access and modify the state dict in Fabric because it gets called internally. That's the motivation for adding this filter hook. The rest of the state  (metadata) is provided by the user, so the user has always full access to it and filtering can be done without special API in Fabric.